snapcraft.yaml

```diff
@@ -159,6 +161,19 @@ parts:
         craftctl default
 
         mkdir -p "${CRAFT_PART_INSTALL}/etc/ovn/"
+    override-stage: |
+        craftctl default
+        pkg_version=$(
+            dpkg-deb -f \
+            $CRAFT_PART_SRC/../stage_packages/ovn-host*.deb \
+            Version | sed -rne 's/([0-9.]+)[-+].*$$/\1/p')
+        git_version=$(
+            git -C $CRAFT_PROJECT_DIR describe \
+                --always \
+                --dirty \
+                --tags \
+                --abbrev=10 | sed -s 's/^v//')
+        craftctl set version=${pkg_version}+${git_version}
```

Comment:
As @freyes pointed out, using the Git sha from a different project than the first part of the version might be confusing and not what the idle observer might expect.

Perhaps we need to prefix it with something, i.e. `22.03.0+microovn6001c4d922`, it becomes a bit long though.

We need to bear in mind that the maximum length of the version string is 32 characters ([relevant discussion](https://forum.snapcraft.io/t/feature-request-extend-length-of-version-string/13335))

Thoughts @dshcherb @mkalcok @sabaini ?

Reply:
Yep, with `microceph` in there it'd be even longer. 

Some options that come to mind

1. Document the semantics of the version string somewhere else so at least ppl can look it up
2. Use some mnemonic, e.g. `22.03.0+mo6001c4d922` for `microovn` 
3. Or alternatively a mnemonic making it clearer the snap git has is referenced `22.03.0+snap6001c4d922`

Nothing's ideal but IMHO 3. might be the least worst of the ones above?

Reply:
Yes, I like alternative 3 as well, it's both shorter and applies equally well to both (and other similar) projects.

Reply:
I'm just thinking out loud here, but do we need to involve version from ovn package at all? We could just add tag to this repository and use
```
git describe --tags --always HEAD
```
 to generate the version.
 Until we add first tag, this command will output just hash which is imo fine, it's just temporary. Then lets say we tag our first release with `22.03`, the command above will output just `22.03` which looks good in the snap store. And then when we push commits after the tag we'll get `<tagname>-<ncommits-ahead>-<commit-id>`

Reply:
Thank you for providing your thoughts @mkalcok!

Assuming we would want to tag the microovn repository following the upstream versions, your suggestion is not unreasonable, and my question then becomes; is that what we want to do?

Snapcraft provides us with the means to declare a set of parts to put into a package. We can either source parts from the Ubuntu distribution and associated archives of debs (such as UCA), or source parts directly from upstream Git repositories.

In both cases we will be able to consume a versioned artifact that has already received release engineering from either of those upstream projects.

The microovn snap is in essence a deployment tool, simpler and still similar to charms. From previous experience with the OVN charms, there is not much change in the deployment specific parts for each upstream release.

My reasoning behind building the version based on the upstream git/package version + some other version for the snap part is that I expect the upstream version to change much more often than the corresponding snap version. If we were to tag the snap project for every upstream version, I think we will be inventing unnecessary busywork for ourselves.

The need for exposing the snap version comes from changes we make to the snap specific code, which may change within the same upstream OVN version, and we may want to know which version of the snap code a user is running whenever an issue is reported.

Reply:
> The microovn snap is in essence a deployment tool, simpler and still similar to charms

although from a QA perspective you would want to lockdown the snap to a specific ovn version, if the deb package moves to a new ovn version you would want to (re)validate the snap when you are rebuilding with this new version, and at that point do the tagging sounds reasonable.

Depending on how your CI/CD pipeline looks like you could do the tagging at that point automagically.

I think in the end we would want to make the conscious decision if the snap is loosely coupled to the ovn version (AKA it has code to adapt itself to the version of ovn it's carrying) or not, for example in the sunbeam world the charm and the rock are strongly coupled

Reply:
> > The microovn snap is in essence a deployment tool, simpler and still similar to charms
> 
> although from a QA perspective you would want to lockdown the snap to a specific ovn version, if the deb package moves to a new ovn version you would want to (re)validate the snap when you are rebuilding with this new version, and at that point do the tagging sounds reasonable.

We definitively want validation for any new major version, and I think that a new major version will be expressed as a new track in the snap store.

Our current pipeline is based on [Launchpad snap recipes](https://launchpad.net/microovn/+snaps), which consumes the code from branches or tags to build a certain track.

Now, if we for example feed a branch into the Launchpad snap recipe machinery, locking the OVN part down to whatever is available in the base Ubuntu distribution defined for that branch, we would be a beneficiary of all the work that goes into that Ubuntu distribution throughout its lifetime, including security and point release updates. Ideally, those kinds of updates can just flow through without manual intervention, at least to the edge channel.

Would you think that additional tagging of the snap git repository be necessary in this context?